torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -367,56 +406,98 @@ def batched_powerSGD_hook(
+    # Reuse P and Q from the previous iteration if possible.
+    # The memory spaces of P and Q need to be (re)allocated at the beginning,
+    # as well as later whenever the buckets are rebuilt during training.
+    if (
+        not state.warm_start
+        or bucket_index not in state.p_memory_dict
+        or state.p_memory_dict[bucket_index].shape
+        != (square_side_length, state.matrix_approximation_rank)
+    ):
+        # If warm-start is disabled, low-rank tensors will be initialized at every step.
+        # Only log this if warm-start to avoid spamming.
+        if state.warm_start:
+            logging.info(
+                "Initializing low-rank tensors P and Q, each of which has a shape of {} x {}.".format(
+                    square_side_length, state.matrix_approximation_rank
+                )
+            )
+
+        def create_low_rank_tensor(fill_random_values, rng):
+            "Returns a low-rank 2D tensor of square_side_length * matrix_approximation_rank."
+            if fill_random_values:
+                with torch.random.fork_rng(devices=[]):
+                    # Fork this RNG to avoid changing the seed globally and affecting the random sampling
+                    # anywhere else in the training.
+                    # The seed makes sure that the initial random values are the same across all the DDP replicas.
+                    # Such seed should differ at every step.
+                    # Since it is very slow to fork RNG state across all the CUDA devices,
+                    # only fork on CPU and then move the generated tensor to the CUDA device.
+                    torch.manual_seed(rng.randint(1_000_000_000))
+                    return torch.randn(
+                        square_side_length,
+                        state.matrix_approximation_rank,
+                        device="cpu",
+                        dtype=input_tensor.dtype,
+                    ).to(device)
+            else:
+                return torch.empty(
                     square_side_length,
                     state.matrix_approximation_rank,
-                    device="cpu",
+                    device=device,
                     dtype=input_tensor.dtype,
-                ).to(device)
-        else:
-            return torch.empty(
-                square_side_length,
-                state.matrix_approximation_rank,
-                device=device,
-                dtype=input_tensor.dtype,
-            )
+                )
 
-    p = create_low_rank_tensor(fill_random_values=False, rng=state.rng)
-    q = create_low_rank_tensor(fill_random_values=True, rng=state.rng)
-    _orthogonalize(q, 0)
+        state.p_memory_dict[bucket_index] = create_low_rank_tensor(
+            fill_random_values=False, rng=state.rng
+        )
+        state.q_memory_dict[bucket_index] = create_low_rank_tensor(
+            fill_random_values=True, rng=state.rng
+        )
+    _orthogonalize(state.q_memory_dict[bucket_index], 0)
```

Comment:
Maybe I'm missing something, but feels like there is a slight difference between batched vs non-batched powersgd. 

in unbatched, we set p_memory_dict and q_memory_dict to torch.empty, and then later assign q.data = torch.rand, orthogonalize, and let p = TQ  for some tensor T. In this batched version, we instead initialize p_memory as empty, and q_memory as rand. We then orthogonalize q and let p = TQ. 

So the end result is the same, but it looks like unbatched version takes an extra step to assign q as empty but then re-assigns it to torch.rand. Can we skip that in the unbatched version?

Reply:
Good observation! The subtle difference is true. This is actually intended behavior:
For the first un-batched version, need to call `torch.empty` to allocate contiguous memory (Qs) first, and then call `torch.rand` to randomize each individual Q. Cannot directly randomize Qs at one time like the batched version.